File: Sesame/src/operation/GraphOperation.java
package operation;

import java.io.IOException;
import java.text.DateFormat;
import java.text.SimpleDateFormat;
import java.util.Date;

import org.openrdf.repository.RepositoryConnection;

import exception.OperationException;
import utility.ResultWriter;

public abstract class GraphOperation {
	private static final String resultDirectory = System.getProperty("user.dir") + System.getProperty("file.separator") + "Result";
	protected String name;
	protected ResultWriter writer;
	protected RepositoryConnection connection;
	
	protected GraphOperation(String name, RepositoryConnection connection) throws OperationException {
		if(name == null || connection == null)
			throw new OperationException("Cannot create graph operation.");
		
		this.name = name;
		this.connection = connection;
		
		DateFormat dateFormat = new SimpleDateFormat("yyyyMMdd-HHmmss");
		Date date = new Date(); 
		try {
			writer = new ResultWriter(resultDirectory + System.getProperty("file.separator") + name, dateFormat.format(date));
		} catch (IOException e) {
			throw new OperationException("Cannot create result writer.");
		}
	}
	
	public void execute() {
		long start = System.currentTimeMillis();
		long p0 = start;
		init();
		System.out.println("\tInitialization: " + (System.currentTimeMillis() - start) + " ms");
		start = System.currentTimeMillis();
		process();
		System.out.println("\tProcess: " + (System.currentTimeMillis() - start) + " ms");
		start = System.currentTimeMillis();
		end();
		System.out.println("\tFinalization: " + (System.currentTimeMillis() - start) + " ms");
		System.out.println("Total: " + (System.currentTimeMillis() - p0) + " ms");
	}
	
	protected abstract void init();
	protected abstract void process();
	protected abstract void end();
	
	public void close() {
		if(writer != null)
			writer.close();
	}
}

File: Jena/operation/GraphOperation.py
'''
Created on Jan 21, 2016

@author: Seokyong Hong
'''
import time
from abc import ABCMeta, abstractmethod
from utility.FileWriter import ResultWriter

class GraphOperation(object):
    __metaclass__ = ABCMeta

    def __init__(self, name, connection):
        self.name = name
        self.connection = connection
    
    @abstractmethod
    def initialize(self):
        raise RuntimeError('The initialize function must be overriden.')
    
    @abstractmethod
    def process(self):
        raise RuntimeError('The process function must be overriden.')
    
    @abstractmethod
    def store(self):
        raise RuntimeError('The store function must be overriden.')
        
    @abstractmethod
    def finalize(self):
        raise RuntimeError('The finalize function must be overriden.')
    
    def execute(self):
        start = time.time()
        
        print '[' + self.name + ']'
        
        fstart = time.time()
        self.initialize()
        print '\tInitialization Time: ' + '%f' % (time.time() - fstart)
        
        fstart = time.time()
        self.process()
        print '\tProcess Time: ' + '%f' % (time.time() - fstart)
        
        self.writer = ResultWriter(self.name)
        self.writer.open()
        fstart = time.time()        
        self.store()
        print '\tStore Time: ' + '%f' % (time.time() - fstart)
        self.writer.close()
        
        fstart = time.time()
        self.finalize()
        print '\tFinalization Time: ' + '%f' % (time.time() - fstart)
        
        print 'Execution Time: ' + '%f' % (time.time() - start)